src/rand.rs

```diff
@@ -248,6 +256,20 @@ mod sysrand_or_urandom {
     }
 }
 
+#[cfg(all(
+    target_os = "linux",
+    feature = "sgx",
+    not(any(feature = "dev_urandom_fallback", feature = "use_heap"))))]
+mod sgxrand {
+    use sgx_trts::trts;
+    use crate::error;
+
+    pub fn fill(dest: &mut [u8]) -> Result<(), error::Unspecified> {
+        trts::rsgx_read_rand(dest).map_err(|_| error::Unspecified)
```

Comment:
Is this really a good idea? I looked at the source code for `sgx_raed_rand` in the Intel SGX SDK. It does feature detection using CPUID to try to detect whether RDRAND is available, and falls back to `rand()` if feature detection says RDRAND isn't available. But, feature detection like that isn't secure in SGX because the `CPUID` has to be executed outside the enclave.

In #738, @akash-fortanix and @jethrogb suggested to just hard-code the use of `RDRAND`, i.e. require `RDRAND` be available within SGX, since in practice it seems like all SGX-capable CPUs include RDRAND, despite Intel's warnings. That seems more secure to me. What do you think?

Importantly, I'd rather avoid having two different implementations for this for the two different SGX targets.

Reply:
It sounds like you're saying that Intel's secure randomness isn't really secure.
Assembly is out of my yard.
@dingelish What do you think?

Reply:
> It sounds like you're saying that Intel's secure randomness isn't really secure.
> Assembly is out of my yard.
> @dingelish What do you think?

I've not dug into the randomness of `RDRAND` much. What I know is that people choose to trust it or not.

Technically the randomness of `RDRAND` [seems good enough](https://eprint.iacr.org/2014/504.pdf).

Reply:
@dingelish Whether or not RDRAND is secure, both of these approaches ultimately use RDRAND, right? That is, the `sgx_read_rand` function uses RDRAND under the hood, right? The main questoin is whether it is better to use the `sgx_read_rand` wrapper or whether it's better to just directly use `RDRAND`, right?

I have to say, I don't understand the Intel SDK source code at all. My concern was already raised as a bug at https://github.com/intel/linux-sgx/issues/111. It is confusing to have two different implementations of the function (if I understand correctly): the one you pointed me to and the one at https://github.com/intel/linux-sgx/blob/master/common/src/sgx_read_rand.cpp. This makes me lean in favor of just using RDRAND directly for all SGX targets. WDYT?

Reply:
Yes everything will ultimately use RDRAND. I believe in foreseeable future this is the only randomness source in SGX.

You can forget about `sgx_read_rand`, a function exported by Intel SGX SDK, because ring does not need to depends on Intel SGX SDK. Simply wrap `RDRAND` or just use intrinsic functions and that'll be 100% fine for SGX enclave.

Yes. There are two implementations of `sgx_read_rand`. One is for the untrusted side, and another is for the trusted side (SGX Enclave). The code snipped he referred to is used in the untrusted side.

Reply:
@dingelish Thanks for the explanation. I agree that we should just more directly use RDRAND and avoid `sgx_read_rand`.